                        Import Variants and Misconceptions

In [1]:
import sys

In [2]:
for key in sorted(sys.modules.keys()):
    print(key)

IPython
IPython.core
IPython.core.alias
IPython.core.application
IPython.core.async_helpers
IPython.core.autocall
IPython.core.builtin_trap
IPython.core.compilerop
IPython.core.completer
IPython.core.completerlib
IPython.core.crashhandler
IPython.core.debugger
IPython.core.display
IPython.core.display_functions
IPython.core.display_trap
IPython.core.displayhook
IPython.core.displaypub
IPython.core.error
IPython.core.events
IPython.core.excolors
IPython.core.extensions
IPython.core.formatters
IPython.core.getipython
IPython.core.history
IPython.core.hooks
IPython.core.inputtransformer2
IPython.core.interactiveshell
IPython.core.latex_symbols
IPython.core.logger
IPython.core.macro
IPython.core.magic
IPython.core.magic_arguments
IPython.core.magics
IPython.core.magics.auto
IPython.core.magics.basic
IPython.core.magics.code
IPython.core.magics.config
IPython.core.magics.display
IPython.core.magics.execution
IPython.core.magics.extension
IPython.core.magics.history
IPython.core.magics.loggi

In [3]:
'cmath' in globals()

False

In [4]:
'cmath' in sys.modules

False

In [5]:
from cmath import exp

In [6]:
'cmath' in globals()

False

In [7]:
'exp' in globals()

True

In [8]:
exp

<function cmath.exp(z, /)>

In [9]:
id(exp)

1536902050976

In [10]:
'cmath' in sys.modules

True

In [11]:
exp(2+2j)

(-3.074932320639359+6.71884969742825j)

In [12]:
cmath = sys.modules['cmath']

In [13]:
'cmath' in globals()

True

In [14]:
cmath.exp(2+2j)

(-3.074932320639359+6.71884969742825j)

In [15]:
cmath.sin(2+2j)

(3.4209548611170133-1.5093064853236156j)

In [16]:
cmath.pi

3.141592653589793

In [17]:
cmath.cos(2+2j)

(-1.5656258353157435-3.2978948363112366j)

In [18]:
from cmath import *

In [19]:
globals()

{'__name__': '__main__',
 '__doc__': 'Automatically created module for IPython interactive environment',
 '__package__': None,
 '__loader__': None,
 '__spec__': None,
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '_ih': ['',
  'import sys',
  'for key in sorted(sys.modules.keys()):\n    print(key)',
  "'cmath' in globals()",
  "'cmath' in sys.modules",
  'from cmath import exp',
  "'cmath' in globals()",
  "'exp' in globals()",
  'exp',
  'id(exp)',
  "'cmath' in sys.modules",
  'exp(2+2j)',
  "cmath = sys.modules['cmath']",
  "'cmath' in globals()",
  'cmath.exp(2+2j)',
  'cmath.sin(2+2j)',
  'cmath.pi',
  'cmath.cos(2+2j)',
  'from cmath import *',
  'globals()'],
 '_oh': {3: False,
  4: False,
  6: False,
  7: True,
  8: <function cmath.exp(z, /)>,
  9: 1536902050976,
  10: True,
  11: (-3.074932320639359+6.71884969742825j),
  13: True,
  14: (-3.074932320639359+6.71884969742825j),
  15: (3.4209548611170133-1.5093064853236156j),
  

In [20]:
sin(2+2j)

(3.4209548611170133-1.5093064853236156j)

In [21]:
from math import *

In [22]:
# globals()

{'__name__': '__main__',
 '__doc__': 'Automatically created module for IPython interactive environment',
 '__package__': None,
 '__loader__': None,
 '__spec__': None,
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '_ih': ['',
  'import sys',
  'import sys',
  'for key in sorted(sys.modules.keys()):\n    print(key)',
  '(cmath) in globals()',
  "'cmath' in globals()",
  "'cmath' in sys.modules",
  'from cmath import exp',
  "'cmath' in globals()",
  "'exp' in globals()",
  'exp',
  'id(exp)',
  "'cmath' in sys.modules",
  'exp(2+2j)',
  "cmath = sys.modules['cmath']",
  "'cmath' in globals()",
  'cmath.exp(2+2j)',
  'cmath.sin(2+2j)',
  'cmath.pi',
  'cmath.cos(2+2j)',
  'from cmath import *',
  'globals()',
  'sin(2+2j)',
  'from math import *',
  'globals()'],
 '_oh': {5: False,
  6: False,
  8: False,
  9: True,
  10: <function cmath.exp(z, /)>,
  11: 139959264753680,
  12: True,
  13: (-3.074932320639359+6.71884969742825j),
  15: True,
  16: (-3.074932320639359+6.71884969742825j),
  17: (3.4209548611170133-1.5093064853236158j),
  18: 3.141592653589793,
  19: (-1.5656258353157435-3.297894836311237j),
  21: {...},
  22: (3.4209548611170133-1.5093064853236158j)},
 '_dh': [PosixPath('/home/rich/Desktop/Mathub/deep-Dive-1/class_notes/modules_packages_namespaces'),
  PosixPath('/home/rich/Desktop/Mathub/deep-Dive-1/class_notes/modules_packages_namespaces')],
 'In': ['',
  'import sys',
  'import sys',
  'for key in sorted(sys.modules.keys()):\n    print(key)',
  '(cmath) in globals()',
  "'cmath' in globals()",
  "'cmath' in sys.modules",
  'from cmath import exp',
  "'cmath' in globals()",
  "'exp' in globals()",
  'exp',
  'id(exp)',
  "'cmath' in sys.modules",
  'exp(2+2j)',
  "cmath = sys.modules['cmath']",
  "'cmath' in globals()",
  'cmath.exp(2+2j)',
  'cmath.sin(2+2j)',
  'cmath.pi',
  'cmath.cos(2+2j)',
  'from cmath import *',
  'globals()',
  'sin(2+2j)',
  'from math import *',
  'globals()'],
 'Out': {5: False,
  6: False,
  8: False,
  9: True,
  10: <function cmath.exp(z, /)>,
  11: 139959264753680,
  12: True,
  13: (-3.074932320639359+6.71884969742825j),
  15: True,
  16: (-3.074932320639359+6.71884969742825j),
  17: (3.4209548611170133-1.5093064853236158j),
  18: 3.141592653589793,
  19: (-1.5656258353157435-3.297894836311237j),
  21: {...},
  22: (3.4209548611170133-1.5093064853236158j)},
 'get_ipython': <bound method InteractiveShell.get_ipython of <ipykernel.zmqshell.ZMQInteractiveShell object at 0x7f4ae87add60>>,
 'exit': <IPython.core.autocall.ZMQExitAutocall at 0x7f4ae87c4ac0>,
 'quit': <IPython.core.autocall.ZMQExitAutocall at 0x7f4ae87c4ac0>,
 '_': (3.4209548611170133-1.5093064853236158j),
 '__': {...},
 '___': (-1.5656258353157435-3.297894836311237j),
 'os': <module 'os' from '/usr/lib/python3.9/os.py'>,
 'sys': <module 'sys' (built-in)>,
 '__vsc_ipynb_file__': '/home/rich/Desktop/Mathub/deep-Dive-1/class_notes/modules_packages_namespaces/import_varianta_misconceptions.ipynb',
 '_i': 'from math import *',
 '_ii': 'sin(2+2j)',
 '_iii': 'globals()',
 '_i1': 'import sys',
 '_i2': 'import sys',
 '_i3': 'for key in sorted(sys.modules.keys()):\n    print(key)',
 'key': 'zope',
 '_i4': '(cmath) in globals()',
 '_i5': "'cmath' in globals()",
 '_5': False,
 '_i6': "'cmath' in sys.modules",
 '_6': False,
 '_i7': 'from cmath import exp',
 'exp': <function math.exp(x, /)>,
 '_i8': "'cmath' in globals()",
 '_8': False,
 '_i9': "'exp' in globals()",
 '_9': True,
 '_i10': 'exp',
 '_10': <function cmath.exp(z, /)>,
 '_i11': 'id(exp)',
 '_11': 139959264753680,
 '_i12': "'cmath' in sys.modules",
 '_12': True,
 '_i13': 'exp(2+2j)',
 '_13': (-3.074932320639359+6.71884969742825j),
 '_i14': "cmath = sys.modules['cmath']",
 'cmath': <module 'cmath' (built-in)>,
 '_i15': "'cmath' in globals()",
 '_15': True,
 '_i16': 'cmath.exp(2+2j)',
 '_16': (-3.074932320639359+6.71884969742825j),
 '_i17': 'cmath.sin(2+2j)',
 '_17': (3.4209548611170133-1.5093064853236158j),
 '_i18': 'cmath.pi',
 '_18': 3.141592653589793,
 '_i19': 'cmath.cos(2+2j)',
 '_19': (-1.5656258353157435-3.297894836311237j),
 '_i20': 'from cmath import *',
 'acos': <function math.acos(x, /)>,
 'acosh': <function math.acosh(x, /)>,
 'asin': <function math.asin(x, /)>,
 'asinh': <function math.asinh(x, /)>,
 'atan': <function math.atan(x, /)>,
 'atanh': <function math.atanh(x, /)>,
 'cos': <function math.cos(x, /)>,
 'cosh': <function math.cosh(x, /)>,
 'isclose': <function math.isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)>,
 'isfinite': <function math.isfinite(x, /)>,
 'isinf': <function math.isinf(x, /)>,
 'isnan': <function math.isnan(x, /)>,
 'log': <function math.log>,
 'log10': <function math.log10(x, /)>,
 'phase': <function cmath.phase(z, /)>,
 'polar': <function cmath.polar(z, /)>,
 'rect': <function cmath.rect(r, phi, /)>,
 'sin': <function math.sin(x, /)>,
 'sinh': <function math.sinh(x, /)>,
 'sqrt': <function math.sqrt(x, /)>,
 'tan': <function math.tan(x, /)>,
 'tanh': <function math.tanh(x, /)>,
 'pi': 3.141592653589793,
 'e': 2.718281828459045,
 'tau': 6.283185307179586,
 'inf': inf,
 'infj': infj,
 'nan': nan,
 'nanj': nanj,
 '_i21': 'globals()',
 '_21': {...},
 '_i22': 'sin(2+2j)',
 '_22': (3.4209548611170133-1.5093064853236158j),
 '_i23': 'from math import *',
 'atan2': <function math.atan2(y, x, /)>,
 'ceil': <function math.ceil(x, /)>,
 'copysign': <function math.copysign(x, y, /)>,
 'degrees': <function math.degrees(x, /)>,
 'dist': <function math.dist(p, q, /)>,
 'erf': <function math.erf(x, /)>,
 'erfc': <function math.erfc(x, /)>,
 'expm1': <function math.expm1(x, /)>,
 'fabs': <function math.fabs(x, /)>,
 'factorial': <function math.factorial(x, /)>,
 'floor': <function math.floor(x, /)>,
 'fmod': <function math.fmod(x, y, /)>,
 'frexp': <function math.frexp(x, /)>,
 'fsum': <function math.fsum(seq, /)>,
 'gamma': <function math.gamma(x, /)>,
 'gcd': <function math.gcd(*integers)>,
 'hypot': <function math.hypot>,
 'isqrt': <function math.isqrt(n, /)>,
 'lcm': <function math.lcm(*integers)>,
 'ldexp': <function math.ldexp(x, i, /)>,
 'lgamma': <function math.lgamma(x, /)>,
 'log1p': <function math.log1p(x, /)>,
 'log2': <function math.log2(x, /)>,
 'modf': <function math.modf(x, /)>,
 'pow': <function math.pow(x, y, /)>,
 'radians': <function math.radians(x, /)>,
 'remainder': <function math.remainder(x, y, /)>,
 'trunc': <function math.trunc(x, /)>,
 'prod': <function math.prod(iterable, /, *, start=1)>,
 'perm': <function math.perm(n, k=None, /)>,
 'comb': <function math.comb(n, k, /)>,
 'nextafter': <function math.nextafter(x, y, /)>,
 'ulp': <function math.ulp(x, /)>,
 '_i24': 'globals()'}

In [23]:
sin(2+2j)

TypeError: can't convert complex to float

In [ ]:
from math import *

{'__name__': '__main__',
 '__doc__': 'Automatically created module for IPython interactive environment',
 '__package__': None,
 '__loader__': None,
 '__spec__': None,
 '__builtin__': <module 'builtins' (built-in)>,
 '__builtins__': <module 'builtins' (built-in)>,
 '_ih': ['',
  'import sys',
  'import sys',
  'for key in sorted(sys.modules.keys()):\n    print(key)',
  '(cmath) in globals()',
  "'cmath' in globals()",
  "'cmath' in sys.modules",
  'from cmath import exp',
  "'cmath' in globals()",
  "'exp' in globals()",
  'exp',
  'id(exp)',
  "'cmath' in sys.modules",
  'exp(2+2j)',
  "cmath = sys.modules['cmath']",
  "'cmath' in globals()",
  'cmath.exp(2+2j)',
  'cmath.sin(2+2j)',
  'cmath.pi',
  'cmath.cos(2+2j)',
  'from cmath import *',
  'globals()',
  'sin(2+2j)',
  'from math import *',
  'globals()',
  '# globals()',
  'sin(2+2j)',
  'from math import *',
  'globals()'],
 '_oh': {5: False,
  6: False,
  8: False,
  9: True,
  10: <function cmath.exp(z, /)>,
  11: 139959264753680,
  12: True,
  13: (-3.074932320639359+6.71884969742825j),
  15: True,
  16: (-3.074932320639359+6.71884969742825j),
  17: (3.4209548611170133-1.5093064853236158j),
  18: 3.141592653589793,
  19: (-1.5656258353157435-3.297894836311237j),
  21: {...},
  22: (3.4209548611170133-1.5093064853236158j),
  24: {...}},
 '_dh': [PosixPath('/home/rich/Desktop/Mathub/deep-Dive-1/class_notes/modules_packages_namespaces'),
  PosixPath('/home/rich/Desktop/Mathub/deep-Dive-1/class_notes/modules_packages_namespaces')],
 'In': ['',
  'import sys',
  'import sys',
  'for key in sorted(sys.modules.keys()):\n    print(key)',
  '(cmath) in globals()',
  "'cmath' in globals()",
  "'cmath' in sys.modules",
  'from cmath import exp',
  "'cmath' in globals()",
  "'exp' in globals()",
  'exp',
  'id(exp)',
  "'cmath' in sys.modules",
  'exp(2+2j)',
  "cmath = sys.modules['cmath']",
  "'cmath' in globals()",
  'cmath.exp(2+2j)',
  'cmath.sin(2+2j)',
  'cmath.pi',
  'cmath.cos(2+2j)',
  'from cmath import *',
  'globals()',
  'sin(2+2j)',
  'from math import *',
  'globals()',
  '# globals()',
  'sin(2+2j)',
  'from math import *',
  'globals()'],
 'Out': {5: False,
  6: False,
  8: False,
  9: True,
  10: <function cmath.exp(z, /)>,
  11: 139959264753680,
  12: True,
  13: (-3.074932320639359+6.71884969742825j),
  15: True,
  16: (-3.074932320639359+6.71884969742825j),
  17: (3.4209548611170133-1.5093064853236158j),
  18: 3.141592653589793,
  19: (-1.5656258353157435-3.297894836311237j),
  21: {...},
  22: (3.4209548611170133-1.5093064853236158j),
  24: {...}},
 'get_ipython': <bound method InteractiveShell.get_ipython of <ipykernel.zmqshell.ZMQInteractiveShell object at 0x7f4ae87add60>>,
 'exit': <IPython.core.autocall.ZMQExitAutocall at 0x7f4ae87c4ac0>,
 'quit': <IPython.core.autocall.ZMQExitAutocall at 0x7f4ae87c4ac0>,
 '_': {...},
 '__': (3.4209548611170133-1.5093064853236158j),
 '___': {...},
 'os': <module 'os' from '/usr/lib/python3.9/os.py'>,
 'sys': <module 'sys' (built-in)>,
 '__vsc_ipynb_file__': '/home/rich/Desktop/Mathub/deep-Dive-1/class_notes/modules_packages_namespaces/import_varianta_misconceptions.ipynb',
 '_i': 'from math import *',
 '_ii': 'sin(2+2j)',
 '_iii': '# globals()',
 '_i1': 'import sys',
 '_i2': 'import sys',
 '_i3': 'for key in sorted(sys.modules.keys()):\n    print(key)',
 'key': 'zope',
 '_i4': '(cmath) in globals()',
 '_i5': "'cmath' in globals()",
 '_5': False,
 '_i6': "'cmath' in sys.modules",
 '_6': False,
 '_i7': 'from cmath import exp',
 'exp': <function math.exp(x, /)>,
 '_i8': "'cmath' in globals()",
 '_8': False,
 '_i9': "'exp' in globals()",
 '_9': True,
 '_i10': 'exp',
 '_10': <function cmath.exp(z, /)>,
 '_i11': 'id(exp)',
 '_11': 139959264753680,
 '_i12': "'cmath' in sys.modules",
 '_12': True,
 '_i13': 'exp(2+2j)',
 '_13': (-3.074932320639359+6.71884969742825j),
 '_i14': "cmath = sys.modules['cmath']",
 'cmath': <module 'cmath' (built-in)>,
 '_i15': "'cmath' in globals()",
 '_15': True,
 '_i16': 'cmath.exp(2+2j)',
 '_16': (-3.074932320639359+6.71884969742825j),
 '_i17': 'cmath.sin(2+2j)',
 '_17': (3.4209548611170133-1.5093064853236158j),
 '_i18': 'cmath.pi',
 '_18': 3.141592653589793,
 '_i19': 'cmath.cos(2+2j)',
 '_19': (-1.5656258353157435-3.297894836311237j),
 '_i20': 'from cmath import *',
 'acos': <function math.acos(x, /)>,
 'acosh': <function math.acosh(x, /)>,
 'asin': <function math.asin(x, /)>,
 'asinh': <function math.asinh(x, /)>,
 'atan': <function math.atan(x, /)>,
 'atanh': <function math.atanh(x, /)>,
 'cos': <function math.cos(x, /)>,
 'cosh': <function math.cosh(x, /)>,
 'isclose': <function math.isclose(a, b, *, rel_tol=1e-09, abs_tol=0.0)>,
 'isfinite': <function math.isfinite(x, /)>,
 'isinf': <function math.isinf(x, /)>,
 'isnan': <function math.isnan(x, /)>,
 'log': <function math.log>,
 'log10': <function math.log10(x, /)>,
 'phase': <function cmath.phase(z, /)>,
 'polar': <function cmath.polar(z, /)>,
 'rect': <function cmath.rect(r, phi, /)>,
 'sin': <function math.sin(x, /)>,
 'sinh': <function math.sinh(x, /)>,
 'sqrt': <function math.sqrt(x, /)>,
 'tan': <function math.tan(x, /)>,
 'tanh': <function math.tanh(x, /)>,
 'pi': 3.141592653589793,
 'e': 2.718281828459045,
 'tau': 6.283185307179586,
 'inf': inf,
 'infj': infj,
 'nan': nan,
 'nanj': nanj,
 '_i21': 'globals()',
 '_21': {...},
 '_i22': 'sin(2+2j)',
 '_22': (3.4209548611170133-1.5093064853236158j),
 '_i23': 'from math import *',
 'atan2': <function math.atan2(y, x, /)>,
 'ceil': <function math.ceil(x, /)>,
 'copysign': <function math.copysign(x, y, /)>,
 'degrees': <function math.degrees(x, /)>,
 'dist': <function math.dist(p, q, /)>,
 'erf': <function math.erf(x, /)>,
 'erfc': <function math.erfc(x, /)>,
 'expm1': <function math.expm1(x, /)>,
 'fabs': <function math.fabs(x, /)>,
 'factorial': <function math.factorial(x, /)>,
 'floor': <function math.floor(x, /)>,
 'fmod': <function math.fmod(x, y, /)>,
 'frexp': <function math.frexp(x, /)>,
 'fsum': <function math.fsum(seq, /)>,
 'gamma': <function math.gamma(x, /)>,
 'gcd': <function math.gcd(*integers)>,
 'hypot': <function math.hypot>,
 'isqrt': <function math.isqrt(n, /)>,
 'lcm': <function math.lcm(*integers)>,
 'ldexp': <function math.ldexp(x, i, /)>,
 'lgamma': <function math.lgamma(x, /)>,
 'log1p': <function math.log1p(x, /)>,
 'log2': <function math.log2(x, /)>,
 'modf': <function math.modf(x, /)>,
 'pow': <function math.pow(x, y, /)>,
 'radians': <function math.radians(x, /)>,
 'remainder': <function math.remainder(x, y, /)>,
 'trunc': <function math.trunc(x, /)>,
 'prod': <function math.prod(iterable, /, *, start=1)>,
 'perm': <function math.perm(n, k=None, /)>,
 'comb': <function math.comb(n, k, /)>,
 'nextafter': <function math.nextafter(x, y, /)>,
 'ulp': <function math.ulp(x, /)>,
 '_i24': 'globals()',
 '_24': {...},
 '_i25': '# globals()',
 '_i26': 'sin(2+2j)',
 '_i27': 'from math import *',
 '_i28': 'globals()'}

In [ ]:
sin(2+2j)

TypeError: can't convert complex to float

In [ ]:
from time import perf_counter

In [ ]:
from cmath import sin as c_sin

In [ ]:
sin

<function math.sin(x, /)>

In [ ]:
from math import sin as c_sin

In [ ]:
c_sin

<function math.sin(x, /)>

In [ ]:
from math import sin as r_sin

In [ ]:
r_sin

<function math.sin(x, /)>

Efficiency

In [ ]:
def my_func(a):
    import math
    return math.sqrt(a)

In [ ]:
from time import perf_counter

In [ ]:
from collections import namedtuple

In [ ]:
Timings = namedtuple('Timings', 'timing_1m timing_2, abs_diff rel_diff_perc')

In [ ]:
def compare_timings(timing1, timing2):
    rel_diff = (timing2 - timing1)/ timing1 * 100
    
    timings = Timings(round(timing1, 1),
                      round(timing2, 1),
                      round(timing2 - timing1, 1),
                      round(rel_diff, 2))
    return timings

In [ ]:
compare_timings(1, 2)

Timings(timing_1m=1, timing_2=2, abs_diff=1, rel_diff_perc=100.0)

In [ ]:
test_repeats = 10_000_000

In [ ]:
#### Timing using fully 2qualified module symbol 

In [ ]:
import math

start = perf_counter()
for _ in range(test_repeats):
    math.sqrt(2)
end = perf_counter()
elapsed_fully_qualified = end - start
print(f'Elapsed: {elapsed_fully_qualified}')

Elapsed: 0.8246239350000906


In [ ]:
### Timming using a directly import symbol nmae

In [ ]:
from math import sqrt

start = perf_counter()
for _ in range(test_repeats):
    sqrt(2)
end = perf_counter()
elapsed_direct_symbol = end - start
print(f'Elapsed: {elapsed_direct_symbol}')

Elapsed: 0.5759787430006327


In [ ]:
compare_timings(elapsed_fully_qualified, elapsed_direct_symbol)

Timings(timing_1m=0.7, timing_2=0.6, abs_diff=-0.1, rel_diff_perc=-13.04)

In [ ]:
###  Timming uing a function wrapper (fully qualified symbol)

In [ ]:
import math

def func():
    math.sqrt(2)

start = perf_counter()
for _ in range(test_repeats):
    func()
end = perf_counter()
elapsed_func_fully_qualified = end - start
print(f'Elapsed: {elapsed_func_fully_qualified}')

Elapsed: 1.2538997080009722


In [ ]:
from math import sqrt

def func():
    sqrt(2)

start = perf_counter()
for _ in range(test_repeats):
    func()
end = perf_counter()
elapsed_func_direct_symbol = end - start
print(f'Elapsed: {elapsed_func_direct_symbol}')

Elapsed: 0.9675852989985287


In [ ]:
compare_timings(elapsed_func_fully_qualified, elapsed_func_direct_symbol)

Timings(timing_1m=1.3, timing_2=1.0, abs_diff=-0.3, rel_diff_perc=-22.83)

In [ ]:
### Nested Imports

In [ ]:
def func():
    import math import sqrt
    math.sqrt(2)

start = perf_counter()
for _ in range(test_repeats):
    func()
end = perf_counter()
elapsed_nested_fully_direct = end - start
print(f'Elapsed: {elapsed_nested_fully_qualified}')

Elapsed: 1.9206396469999163


In [ ]:
compare_timings(elapsed_func_fully_qualified, elapsed_nested_fully_qualified)

Timings(timing_1m=1.3, timing_2=1.9, abs_diff=0.7, rel_diff_perc=53.17)

In [ ]:
def func():
    import math
    math.sqrt(2)

start = perf_counter()
for _ in range(test_repeats):
    func()
end = perf_counter()
elapsed_nested_direct_symbol = end - start
print(f'Elapsed: {elapsed_nested_direct_symbol}')

Elapsed: 2.0497856729998603


In [ ]:
compare_timings(elapsed_nested_fully_qualified, elapsed_nested_direct_symbol)

Timings(timing_1m=1.9, timing_2=2.0, abs_diff=0.1, rel_diff_perc=6.72)